### PR TITLE
Add Profile guided optimization support for microsoft.ui.xaml.dll

### PR DESCRIPTION
**Pull request by ranjeshj**
This is a work in progress and still needs a few missing pieces, but starting a PR to get feedback while I figure out the rest. I do not expect to merge this in its current form.

Profile guided optimization is a way to optimize the binary by running scenarios on an instrumented product binary and collecting trace information and feeding that back to the build to do better placement of pages to improve load time, better in-lining etc. More info here if you are curious about what this is - https://www.youtube.com/watch?v=zEsdBcu4R00 

The steps involved are
1.  Instrument the binary and run release tests on the instrumented binary. This will be done in a new pipeline that is similar to the CI pipeline. The release tests are used as scenarios. After each test completes, we run a tool to collect profile counts (pgc files).
2. The pgc files are merged into a profile guided database file (pgd), which is then published as a nuget package.
3. Then we create an optimized build using the pgd file in a new build.

Some steps that are currently missing
1. There is currently no easy way to get the pgc files from the helix test machines back into the pipeline to create the merged pgd file. Helix folks are aware of this and have work items to solve this. In the mean time, I have copied them over manually to my dev box and created a nuget package containing the profile database.

2. Where the profile database nuget package should be published. Currently, i can do the optimized build locally, but we need to figure out where the nuget package will be pushed and if we should allow external developers to be able to reach that or if this should be an internal solution.

3. arm/arm64 data is currently not available because we don't have helix runs for those platforms yet.

### CODE REVIEW

**Comment by ranjeshj** [comment]
> CodeFlow for GitHub does not download executable files. If you trust this file and would like to run it, please download it yourself through GitHub.

this folder should be removed

---
Refers to: build/PGO/Microsoft.UI.Xaml.pgd:1 in 8d7671b. [](commit_id = 8d7671bd34161b722553f94f0d89bc8d25f5e787, deletion_comment = False)

**Comment by Austin-Lamb** [comment]
Are there any measurements yet on the perf impact?  Example things that would be interesting:

1) Change to binary size
2) CPU Impact to some scenario (describe the scenario)
3) Reference Set impact to some scenario as PGO should improve code locality (again, describe the scenario)

**Comment by ranjeshj** [comment]
Don't have any yet. I just built my first mux.dll that is optimized with data from the existing release tests (although just a few tests now). 

---
In reply to: [508267169](https://github.com/microsoft/microsoft-ui-xaml/pull/988#issuecomment-508267169) [](ancestors = 508267169)

File: tools/MUXPGODatabase/NuSpecs/PublishNupkg.cmd [hunk 1]
```diff
@@ -0,0 +1,13 @@
+@ECHO OFF
+SETLOCAL
+set ERRORLEVEL=
+
+for %%A IN (%~dp0MUXPGODatabase.??.??.*.nupkg) do (
+    echo Candidate nuget package: %%A
+    echo %~dp0..\..\NugetWrapper.cmd push %%A -Source https://pkgs.dev.azure.com/ms/_packaging/MUX-CI/nuget/v3/index.json -apikey VSTS
+rem you need to get an authorized version of nuget.exe from https://dev.azure.com/ms/microsoft-ui-xaml/_packaging?_a=package&feed=MUX-CI&package=Microsoft.UI.Xaml&protocolType=NuGet&version=2.2.190702001-prerelease
+    "C:\Users\ranjeshj\Desktop\CredentialProviderBundle\NuGet.exe" push %%A -Source https://pkgs.dev.azure.com/ms/_packaging/MUX-CI/nuget/v3/index.json -apikey VSTS
```
Comment (ranjeshj) [inline tools/MUXPGODatabase/NuSpecs/PublishNupkg.cmd]: >"C:\Users\ranjeshj\Desktop\CredentialProviderBundle\NuGet.exe [](start = 4, length = 61)

This is hacky and should go away once we figure out where the nuget package goes.

**Comment by Austin-Lamb** [comment]
ok - you can also share the locally-built-with-PGO binary and PDB with me, I can look to see if it seems to be working correctly

---
In reply to: [508267598](https://github.com/microsoft/microsoft-ui-xaml/pull/988#issuecomment-508267598) [](ancestors = 508267598,508267169)

File: mux.PGO.props [hunk 2]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+    <PropertyGroup>
+        <PlatformShortName Condition="'$(PlatformShortName)'==''">$(Platform)</PlatformShortName>
+        <PGOFilesDir>$(VCToolsInstallDir)bin\Host$(PlatformShortName)\$(PlatformShortName)</PGOFilesDir>
+        <PGOFilesDir Condition="'$(PlatformShortName)'=='ARM' or '$(PlatformShortName)'=='ARM64'">$(VCToolsInstallDir)bin\$(PlatformShortName)</PGOFilesDir>       
+    </PropertyGroup>
+    <PropertyGroup>    
+        <PGOBuildMode Condition="'$(PGOBuildMode)'==''">Optimize</PGOBuildMode>
```
Comment (ranjeshj) [inline mux.PGO.props]: >Optimize [](start = 56, length = 8)

we could probably do with not having this option and always optimize if the profile guided database is available.

Comment (ranjeshj) [reply mux.PGO.props]: This seems useful when building locally to quickly enable/disable optimizations. So decided to keep it.

---
In reply to: [300175577](https://github.com/microsoft/microsoft-ui-xaml/pull/988#discussion_r300175577) [](ancestors = 300175577)

File: tools/MUXPGODatabase/NuSpecs/PublishNupkg.cmd [hunk 3]
```diff
@@ -4,9 +4,9 @@ set ERRORLEVEL=
 
 for %%A IN (%~dp0MUXPGODatabase.??.??.*.nupkg) do (
     echo Candidate nuget package: %%A
-    echo %~dp0..\..\NugetWrapper.cmd push %%A -Source https://pkgs.dev.azure.com/ms/_packaging/MUX-CI/nuget/v3/index.json -apikey VSTS
-rem you need to get an authorized version of nuget.exe from https://dev.azure.com/ms/microsoft-ui-xaml/_packaging?_a=package&feed=MUX-CI&package=Microsoft.UI.Xaml&protocolType=NuGet&version=2.2.190702001-prerelease
-    "C:\Users\ranjeshj\Desktop\CredentialProviderBundle\NuGet.exe" push %%A -Source https://pkgs.dev.azure.com/ms/_packaging/MUX-CI/nuget/v3/index.json -apikey VSTS
+    echo %~dp0..\..\NugetWrapper.cmd push %%A -Source https://microsoft.pkgs.visualstudio.com/DefaultCollection/_packaging/DEPControls/nuget/v3/index.json -apikey VSTS
+rem you need to get an authorized version of nuget.exe from https://dev.azure.com/microsoft/WinUI/_packaging?_a=feed&feed=DEPControls
+    "T:\WinUICredential\NuGet.exe" push %%A -Source https://microsoft.pkgs.visualstudio.com/DefaultCollection/_packaging/DEPControls/nuget/v3/index.json -apikey VSTS
```
Comment (jevansaks) [inline tools/MUXPGODatabase/NuSpecs/PublishNupkg.cmd]: Use tools\NugetWrapper.cmd so this script can work locally?

File: build/Helix/PrepareHelixPayload.ps1 [hunk 4]
```diff
@@ -50,6 +50,8 @@ Copy-If-Exists "$repoDirectory\Artifacts\$ArtifactName\$Configuration\$Platform\
 
 # NuGet test artifacts
```
Comment (kmahone) [inline build/Helix/PrepareHelixPayload.ps1]: Please update this comment to something like
# Copy files from the 'NugetPkgTestsDrop' or 'FrameworkPkgTestsDrop' artifact dir

File: build/MUX-PGOInstrument.yml [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  strategy:
+    maxParallel: 10
+    matrix:
+      Release_x86:
+        buildPlatform: 'x86'
+        buildConfiguration: 'Release'
+      Release_x64:
+        buildPlatform: 'x64'
+        buildConfiguration: 'Release'
+      Release_Arm:
```
Comment (kmahone) [inline build/MUX-PGOInstrument.yml]: Are we doing PGO for arm/arm64? If not, these are unnecessary.

File: test/MUXControlsReleaseTest/MUXControls.ReleaseTest/NugetTestsCX.cs [hunk 6]
```diff
@@ -29,6 +29,12 @@ public static void ClassInitialize(TestContext testContext)
             TestEnvironment.Initialize(testContext, TestType.NugetCX);
         }
 
+        [TestCleanup]
+        public void TestCleanup()
+        {
+            PGOManager.PGOSweepIfInstrumented(TestEnvironment.TestContext.TestName);
```
Comment (kmahone) [inline test/MUXControlsReleaseTest/MUXControls.ReleaseTest/NugetTestsCX.cs]: Do we need to sweep for every test method, or can we do it all at once in AssemblyCleanup?

File: test/MUXControlsReleaseTest/MUXControls.ReleaseTest/PGOManager.cs [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+                LogOutput("Failed trying to pgosweep. " + ex.ToString());
+                throw;
+            }
+#endif
+        }
+
+        public static void LogOutput(string message)
+        {
+#if USING_TAEF
+            Log.Comment(message);
```
Comment (kmahone) [inline test/MUXControlsReleaseTest/MUXControls.ReleaseTest/PGOManager.cs]: Why is this necessary? Most of the other tests just call Log.Comment, which will resolve to either TAEF or VSTest depending on the usings.

File: test/MUXControlsReleaseTest/NugetPackageTestAppCX/NugetPackageTestAppCX.vcxproj [hunk 8]
```diff
@@ -194,6 +195,27 @@
...
+      <Visible>false</Visible>
+      <InProject>false</InProject>
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
+	  <None Include="$(PGOFilesDir)\pgosweep.exe">
+      <Visible>false</Visible>
+      <InProject>false</InProject>
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
+	  <None Include="$(PGOFilesDir)\vcruntime140.dll" Condition="Exists('$(PGOFilesDir)\vcruntime140.dll')">
```
Comment (kmahone) [inline test/MUXControlsReleaseTest/NugetPackageTestAppCX/NugetPackageTestAppCX.vcxproj]: indent

File: tools/MUXPGODatabase/NuSpecs/MUXPGODatabase.nuspec [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+  <metadata>
+    <id>MUXPGODatabase</id>
+    <version>0.0.1</version>
+    <title>MUXPGODatabase</title>
+    <authors>Microsoft</authors>
+    <owners>Microsoft</owners>
+    <requireLicenseAcceptance>false</requireLicenseAcceptance>
+    <description>profile guided optimization database obtained after running training scenarios.</description>
+    <releaseNotes>Initial release.</releaseNotes>
+    <copyright>Copyright 2017</copyright>
```
Comment (kmahone) [inline tools/MUXPGODatabase/NuSpecs/MUXPGODatabase.nuspec]: 2019

File: tools/MUXPGODatabase/NuSpecs/MUXPGODatabase.nuspec [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+  </metadata>
+  <files>
+    <file target="build\MUXPGO.targets" src="build\MUXPGO.targets"/>
+    <file target="tools\x86\microsoft.ui.xaml.pgd" src="tools\x86\microsoft.ui.xaml.pgd"/>
+	  <file target="tools\x64\microsoft.ui.xaml.pgd" src="tools\x64\microsoft.ui.xaml.pgd"/>
```
Comment (kmahone) [inline tools/MUXPGODatabase/NuSpecs/MUXPGODatabase.nuspec]: indentation

File: tools/MUXPGODatabase/NuSpecs/MUXPGODatabase.nuspec [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+    <title>MUXPGODatabase</title>
+    <authors>Microsoft</authors>
+    <owners>Microsoft</owners>
+    <requireLicenseAcceptance>false</requireLicenseAcceptance>
+    <description>profile guided optimization database obtained after running training scenarios.</description>
+    <releaseNotes>Initial release.</releaseNotes>
+    <copyright>Copyright 2017</copyright>
+  </metadata>
+  <files>
+    <file target="build\MUXPGO.targets" src="build\MUXPGO.targets"/>
```
Comment (kmahone) [inline tools/MUXPGODatabase/NuSpecs/MUXPGODatabase.nuspec]: Convention is that the .targets file is named the same as the nupkg (i.e. MUXPGODatabase.targets)

**Comment by kmahone** [comment]
> <?xml version="1.0" encoding="utf-8"?>

Nit: Convention is that a .targets file includes msbuild Targets. So this should be a .props file.

---
Refers to: tools/MUXPGODatabase/NuSpecs/build/MUXPGO.targets:1 in 7389068. [](commit_id = 7389068f475161a23b6c30cfc0f5c58ba7cccc62, deletion_comment = False)

File: build/AzurePipelinesTemplates/MUX-BuildProject-Steps.yml [hunk 12]
```diff
@@ -38,7 +48,7 @@ steps:
       vsVersion: 15.0
       platform: '$(buildPlatform)'
       configuration: '$(buildConfiguration)'
-      msbuildArgs: '/p:AppxPackageDir="${{ parameters.appxPackageDir }}" /p:AppxBundle=Never /p:AppxSymbolPackageEnabled=false /binaryLogger:$(Build.SourcesDirectory)/${{ parameters.solutionPath }}.$(buildPlatform).$(buildConfiguration).binlog -property:MUXVersionBuild=$(builddate_yymm) -property:MUXVersionRevision=$(builddate_dd)$(buildrevision)'
+      msbuildArgs: '/p:AppxPackageDir="${{ parameters.appxPackageDir }}" /p:AppxBundle=Never /p:AppxSymbolPackageEnabled=false /binaryLogger:$(Build.SourcesDirectory)/${{ parameters.solutionPath }}.$(buildPlatform).$(buildConfiguration).binlog -property:MUXVersionBuild=$(builddate_yymm) -property:MUXVersionRevision=$(builddate_dd)$(buildrevision) /p:VCToolsInstallDir="$(VCToolsInstallDir)\" -p:PGOBuildMode=$(PGOBuildMode)'
```
Comment (kmahone) [inline build/AzurePipelinesTemplates/MUX-BuildProject-Steps.yml]: We have a mix of /property: /p: and -p: switches. We should be consistent.

Comment (kmahone) [inline build/AzurePipelinesTemplates/MUX-BuildProject-Steps.yml]: Is is possible to split this line across multiple lines in yml? I don't know if yml supports that, but it would aid readability. This line is becoming a beast!

File: build/PGO/Microsoft.UI.Xaml.PGO.targets [hunk 13]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
```
Comment (kmahone) [inline build/PGO/Microsoft.UI.Xaml.PGO.targets]: Consider renaming to .props.

File: mux.PGO.props [hunk 14]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+    <PropertyGroup>
+        <PlatformShortName Condition="'$(PlatformShortName)'==''">$(Platform)</PlatformShortName>
```
Comment (kmahone) [inline mux.PGO.props]: What is the purpose of 'PlatformShortName' over 'Platform'?

File: dev/dll/Microsoft.UI.Xaml.vcxproj [hunk 15]
```diff
@@ -697,4 +697,5 @@
     <!-- Tell CreateLCGFile to use the LCE file from RC.EXE -->
     <LCEFile>$(_LceFile)</LCEFile>
   </PropertyGroup>
+  <Import Project="$(MSBuildProjectDirectory)\..\..\Build\PGO\Microsoft.UI.Xaml.PGO.targets" />
```
Comment (kmahone) [inline dev/dll/Microsoft.UI.Xaml.vcxproj]: Is there a reason that this is at the end, instead of at the top with the rest of the Imports?

File: mux.PGO.props [hunk 16]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (kmahone) [inline mux.PGO.props]: Do we really need a new top level directory in the repo just for this one file?

File: test/MUXControlsTestApp/MUXControlsTestApp.Shared.shproj [hunk 17]
```diff
@@ -4,11 +4,10 @@
   <PropertyGroup Label="Globals">
     <ProjectGuid>{6AA772A6-CBF7-4FF3-8864-BC9366015DC2}</ProjectGuid>
     <MinimumVisualStudioVersion>15.0</MinimumVisualStudioVersion>
-  </PropertyGroup>
+  </PropertyGroup>  
   <Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props" Condition="Exists('$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props')" />
   <Import Project="$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\CodeSharing\Microsoft.CodeSharing.Common.Default.props" />
   <Import Project="$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\CodeSharing\Microsoft.CodeSharing.Common.props" />
-  <PropertyGroup />
```
Comment (kmahone) [inline test/MUXControlsTestApp/MUXControlsTestApp.Shared.shproj]: Wow. I am surprised we were even able to build with this error!

Comment (ranjeshj) [reply test/MUXControlsTestApp/MUXControlsTestApp.Shared.shproj]: That is just considered as an empty PropertyGroup tag.

---
In reply to: [301743667](https://github.com/microsoft/microsoft-ui-xaml/pull/988#discussion_r301743667) [](ancestors = 301743667)

File: test/MUXControlsReleaseTest/NugetPackageTestAppCX/NugetPackageTestAppCX.vcxproj (hunk 8 above)
Comment (kmahone) [inline test/MUXControlsReleaseTest/NugetPackageTestAppCX/NugetPackageTestAppCX.vcxproj]: I am surprised that this is necessary, since the app uses the VCLibs framework package. Do you know why this is necessary?

File: build/MUX-PGOInstrument.yml (hunk 5 above)
Comment (ranjeshj) [reply build/MUX-PGOInstrument.yml]: The nuget package building step seems to require this. We need to add tests for arm as well shortly, so leaving this for now.

---
In reply to: [301732414](https://github.com/microsoft/microsoft-ui-xaml/pull/988#discussion_r301732414) [](ancestors = 301732414)

File: test/MUXControlsReleaseTest/MUXControls.ReleaseTest/NugetTestsCX.cs (hunk 6 above)
Comment (ranjeshj) [reply test/MUXControlsReleaseTest/MUXControls.ReleaseTest/NugetTestsCX.cs]: I like to keep this separate and the count files smaller. 

---
In reply to: [301736142](https://github.com/microsoft/microsoft-ui-xaml/pull/988#discussion_r301736142) [](ancestors = 301736142)

**Comment by ranjeshj** [comment]
> <?xml version="1.0" encoding="utf-8"?>

I'm not sure if .props file will be automatically included when the nuget package is installed. 

---
In reply to: [509761548](https://github.com/microsoft/microsoft-ui-xaml/pull/988#issuecomment-509761548) [](ancestors = 509761548)

---
Refers to: tools/MUXPGODatabase/NuSpecs/build/MUXPGO.targets:1 in 7389068. [](commit_id = 7389068f475161a23b6c30cfc0f5c58ba7cccc62, deletion_comment = False)

File: test/MUXControlsReleaseTest/NugetPackageTestAppCX/NugetPackageTestAppCX.vcxproj (hunk 8 above)
Comment (ranjeshj) [reply test/MUXControlsReleaseTest/NugetPackageTestAppCX/NugetPackageTestAppCX.vcxproj]: the tools seem to need these in the same folder.

---
In reply to: [301745225](https://github.com/microsoft/microsoft-ui-xaml/pull/988#discussion_r301745225) [](ancestors = 301745225)

File: mux.PGO.props (hunk 16 above)
Comment (ranjeshj) [reply mux.PGO.props]: There is no top level directory, just a file in the top level.

---
In reply to: [301743348](https://github.com/microsoft/microsoft-ui-xaml/pull/988#discussion_r301743348) [](ancestors = 301743348)

File: mux.PGO.props (hunk 14 above)
Comment (ranjeshj) [reply mux.PGO.props]: Win32 vs x86

---
In reply to: [301742511](https://github.com/microsoft/microsoft-ui-xaml/pull/988#discussion_r301742511) [](ancestors = 301742511)

File: mux.PGO.props (hunk 16 above)
Comment (kmahone) [reply mux.PGO.props]: Weird. Why is CodeFlow showing this as a file in a folder named "mux.PGO.props"? 

---
In reply to: [302109935](https://github.com/microsoft/microsoft-ui-xaml/pull/988#discussion_r302109935) [](ancestors = 302109935,301743348)
